docs/dev/rfcs/query/0007-semi-structured-data-types.md

```diff
@@ -0,0 +1,106 @@
+For example, MongoDB and PostgreSQL use [BSON](https://bsonspec.org/) and [jsonb](https://www.postgresql.org/docs/14/datatype-json.html) respectively to store data in JSON format.
+[UBJSON](https://ubjson.org/) is also a compatible format specification for binary JSON, it can provide universal compatibility, as easy of use as JSON while being faster and more efficient.
+All of these binary JSON formats have better performance, the only problem is they lack a good Rust implementation libraries.
+
+#### Store each unique key of data in sub-columns
+
+Despite JSON format can represent arbitrary data, in practice, JSON data is usually generated by the machine, so we can predict the Shema and structure.
+Based on this feature, we can extract and flatten each unique key in JSON data into multiple independent virtual sub-columns.
+
+For example, suppose we have a column named `tweet` and store the following JSON data:
+
+```json
+{"id":1, "date": "1/11", type: "story", "score": 3, "desc": 2, "title": "...", "url": "..."}
+{"id":2, "date": "1/12", type: "poll", "score": 5, "desc": 2, "title": "..."}
+{"id":3, "date": "1/13", type: "pollop", "score": 6, "poll": 2, "title": "..."}
+{"id":4, "date": "1/14", type: "story", "score": 1, "desc": 1, "title": "...", "url": "..."}
+{"id":5, "date": "1/15", type: "comment", "parent": 4, "text": "..."}
+{"id":6, "date": "1/16", type: "comment", "parent": 1, "text": "..."}
+{"id":7, "date": "1/17", type: "pollop", "score": 3, "poll": 2, "title": "..."}
+{"id":8, "date": "1/18", type: "comment", "parent": 1, "text": "..."}
+```
+
+This column can be split into 10 virtual sub-columns: `tweet.id`, `tweet.date`, `tweet.type`, `tweet.score`, `tweet.desc`, `tweet.title`, `tweet.url`, `tweet.parent`, `tweet.text`, `tweet.poll`.
+The data type of each sub-column can also be automatically deducted from the value, then we can automatically create those sub-columns and insert the corresponding value.
+
+The main advantage of this storage format is that it does not need to parse the raw JSON string when querying the data, which can greatly speed up the query processing.
+The disadvantage is that additional processing is required when inserting data, and the schema of data in each row is not exactly the same. In some scenarios with large differences, many sub-column data will be Null.
+In order to have good performance and balance in various scenarios, we can refer to the optimization algorithms introduced in the paper [JSON Tiles](https://db.in.tum.de/people/sites/durner/papers/json-tiles-sigmod21.pdf).
+
+From the perspective of performance, a better solution is to store data in binary JSON-like format and extract some frequently queried unique keys as sub-columns.
+However, in order to simplify development, we use the JSON format in the first version.
+Binary JSON-like format and separately stored sub-columns will be adopted in a future optimized version.
```

Comment:
Does it mean that for the first version, those cols will not be stored in a "column-oriented" way?

how about the [dremel](https://github.com/julienledem/redelm/wiki/The-striping-and-assembly-algorithms-from-the-Dremel-paper) style encoding of nested data structure (which is supported by parquet out of the box)

Reply:
That's pretty complicated, and it seems to only work for the fixed schema column.

Reply:
parquet nested data type must define schema, not suitable for storing such arbitrarily data.